### PR TITLE
Bidirectional A*

### PR DESCRIPTION
**Pull request by SChudakov**
Implemented bidirectional version of the A* shortest path algorithm discussed in #739. 
For the implementation the article Ira Sheldon Pohl, Bi-Directional and Heuristic Search in Path Problems. Ph.D. Dissertation. Stanford University, Stanford, CA, USA. AAI7001588 was used. The algorithm itself described  on pages 95-96.

----
 
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

**Comment by SChudakov** [comment]
I have also added benchmarks to the to compare it with other algorithms in the library. Here are the results:

---------------------------------------------------------------------------------------
|Averaging results over 10 executions                      |setup (ms)| execution (ms)|
|----------------------------------------------------------|----------|---------------|
|Dijkstra with ClosestFirstIterator                        |  7.200   |    22072.100  |
|Dijkstra                                                  |  7.300   |    14204.500  |
|A* no heuristic                                           |  6.000   |    20331.100  |
|A* with ALT heuristic (1 random landmarks)                |  6.400   |    25057.300  |
|A* with ALT heuristic (5 random landmarks)                |  6.200   |    26987.900  |
|Bidirectional Dijkstra                                    |  6.100   |    6180.400   |
|Bidirectional A* with ALT heuristic (1 random landmarks)  |  6.700   |    24753.400  |
|Bidirectional A* with ALT heuristic (5 random landmarks)  |  6.800   |    25950.500  |
|Bidirectional A* no heuristic                             |  6.900   |    38260.200  |
---------------------------------------------------------------------------------------

**Comment by SChudakov** [comment]
For benchmarks I use `DijkstraShortestPathPerformanceTest` where the time is averaged over 10 iteration.
Here is what I get currently in the master branch:

All-Pairs Shortest Paths Benchmark
---------
Using G(n,p) random graph with n = 250, p = 0.3
Warmup phase 5 executions
Averaging results over 10 executions
-------------------------------------------------------------------------
|Averaging results over 10 executions      |setup (ms)  | execution (ms)|
|------------------------------------------|------------|---------------|
Dijkstra with ClosestFirstIterator         | 6.600 (ms) | 23535.900     |
Dijkstra                                   | 6.600 (ms) | 14058.200     |
A* no heuristic                            | 6.800 (ms) | 20384.400     |
A* with ALT heuristic (1 random landmarks) | 7.200 (ms) | 26547.700     |
A* with ALT heuristic (5 random landmarks) | 6.800 (ms) | 27091.500     |
Bidirectional Dijkstra                     | 6.200 (ms) | 6667.900      |
-------------------------------------------------------------------------

**Comment by jkinable** [comment]
The performance results seem to be a bit problematic: the bidirectional approach seems very comparable, or slower than the onedirectional approach? Is there anything that could be done to improve the performance? Usually bidir approaches are more efficient.

**Comment by SChudakov** [comment]
I have some doubts about how to implement what I have described. If a single stopping condition is used throughout the algorithm then it is bounded to using only consistent heuristics. Another option is to adjust the algorithm depending of what type of heuristic provided. In this case it is needed to create some utility functions to determine the type of a heuristic. This can be done by adding a 

`HeuristicType getType(Graph<V,E> graph)`

method to the AStarAdmissibleHeuristic class with the default implementation which will determine the heuristic type using definitions. Consequently this method can be overridden in the subclasses of the interface. For now the HeuristicType enum will consist of ADMISSIBLE, FEASIBLE and CONSISTENT values. @d-michail, @jkinable what is your opinion on this issue?

**Comment by d-michail** [comment]
Since our interface is called `AStarAdmissibleHeuristic`, I think it suffices to add a method 
```
boolean isConsistent();
```
and return false by default. 

In your suggestion you also include the graph as a parameter. Does the type of the heuristic depend on actual graph instance?

**Comment by SChudakov** [comment]
For such heuristics as Euclidean distance or Great Circle distance the consistency does not depend on the graph. But originally I thought of such cases when the heuristic values are specified for a specific pair of vertices, like it is done in AStarShortestPathTest. For example when a heuristic is specified for the graph with Integer vertices it might happen that for some graphs it is consistent and for others  it is not. Consider a following example:
graph 1:
vertices {1,2,3}
edges:
        (1,2) with weight 1
        (2,3) with weight 1
        (1,3) with weight 3
graph 2:
vertices {1,2,3}
edges:
        (1,2) with weight 1
        (2,3) with weight 1
        (1,3) with weight 4
(both graphs are undirected)
heuristic:
  (1,3) -> 4
  (3,1) -> 4
  else 0
For second graph the heuristic is consistent and for the first it is not even admissible. Therefore for such case it depends on the graph. So I thought it would be nice if there were a default implementation of the method which  checks the triangle inequality condition for every edge in the graph.
Concerning feasibility. I think that this trasformation from to consistent heuristic using forward and reversed estimate values can be done simply by defining another heuristic class which reuses the feasible one. So there is no necessity to provide analogous function for fisibility. 
Does that what I have described here make sense for you? Anyway would be glad to get feedback on this topic.

**Comment by SChudakov** [comment]
For such heuristics as Euclidean distance or Great Circle distance the consistency does not depend on the graph. But originally I thought of such cases when the heuristic values are specified for a specific pair of vertices, like it is done in AStarShortestPathTest. For example when a heuristic is specified for the graph with Integer vertices it might happen that for some graphs it is consistent and for others it is not. Consider a following example:
graph 1:
vertices {1,2,3}
edges:
(1,2) with weight 1
(2,3) with weight 1
(1,3) with weight 3
graph 2:
vertices {1,2,3}
edges:
(1,2) with weight 2
(2,3) with weight 2
(1,3) with weight 4
(both graphs are directed)
heuristic:
(1,3) -> 4
else 0
For second graph the heuristic is consistent and for the first it is not even admissible. Therefore for such case it depends on the graph. So I thought it would be nice if there were a default implementation of the method which checks the triangle inequality condition for every edge in the graph.
Concerning feasibility. I think that this trasformation from to consistent heuristic using forward and reversed estimate values can be done simply by defining another heuristic class which reuses the feasible one. So there is no necessity to provide analogous function for fisibility.
Does that what I have described here make sense for you? Anyway would be glad to get feedback on this topic.

**Comment by d-michail** [comment]
I just noticed that we already have a method `AStarShortestPath#isConsistentHeuristic()` which verifies by checking monotonicity for each edge. 

Maybe we can move that implementation in the interface and call it something like `boolean isConsistent(Graph<V,E>)`. Then heuristic implementors can always override and return true if they know that their heuristic is monotone.

**Comment by SChudakov** [comment]
I have done the suggested improvements.

**Comment by SChudakov** [comment]
I have done the suggested improvements.

**Comment by d-michail** [comment]
We follow a 2-version deprecation policy. You should deprecate the old method but keep it. We will be able to remove it after the next release.

**Comment by SChudakov** [comment]
Done.

**Comment by SChudakov** [comment]
Done.

**Comment by d-michail** [comment]
Cool. Please try to keep the authorship consistent when moving code around.

**Comment by SChudakov** [comment]
Done.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ALTAdmissibleHeuristic.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * (C) Copyright 2016-2018, by Dimitrios Michail and Contributors.
+ * (C) Copyright 2016-2018, by Dimitrios Michail, Semen Chudakov and Contributors.
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ALTAdmissibleHeuristic.java]: Please keep only the main author here. The rest are as "Contributors". Git blame keeps the exact logs.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BaseBidirectionalShortestPathAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
+/*
+ * (C) Copyright 2019-2019, by Semen Chudakov and Contributors.
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BaseBidirectionalShortestPathAlgorithm.java]: This code was copied pasted from another file and you changed authorship. Please do not do that.

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BidirectionalAStarShortestPath.java [hunk 3]
```diff
@@ -0,0 +1,373 @@
+/*
+ * (C) Copyright 2019-2019, by Semen Chudakov, Dimitrios Michail, Joris Kinable, Jon Robison, Thomas Breitbart and Contributors.
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BidirectionalAStarShortestPath.java]: Keep only the first name here. 

In general the following rule applied. In the header we only list the main author (some exceptions are probably in the codebase) and we list additional authors in the main class javadocs.

**Comment by SChudakov** [comment]
Done,

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BaseBidirectionalShortestPathAlgorithm.java (hunk 2 above)
Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BaseBidirectionalShortestPathAlgorithm.java]: Still needs to change to the original author. Refactoring techniques such as "extract supertype" do not change the author of the code.

Comment (SChudakov) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BaseBidirectionalShortestPathAlgorithm.java]: Done.

**Comment by SChudakov** [comment]
I am interested in finishing this PR. Are there any other steps I need to take in order for it to be merged?

**Comment by jsichi** [comment]
@d-michail @jkinable do you have followup comments?

**Comment by d-michail** [comment]
No more comments. Looks good.

**Comment by d-michail** [comment]
Thanks!

**Pull request merged**